op-service/sources/prefetching_eth_client.go

```diff
@@ -0,0 +1,113 @@
+package sources
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-service/client"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/sources/caching"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type PrefetchingEthClient struct {
+	inner *EthClient
+	// state that manages pre-fetching, like in-flight fetching jobs
+}
+
+// NewPrefetchingEthClient returns a [PrefetchingEthClient], wrapping an RPC with bindings to fetch ethereum data with added error logging,
+// metric tracking, and caching. The [PrefetchingEthClient] uses concurrent RPC requests to keep a cache of data ready for callers.
+func NewPrefetchingEthClient(client client.RPC, log log.Logger, metrics caching.Metrics, config *EthClientConfig) (*PrefetchingEthClient, error) {
+	if err := config.Check(); err != nil {
+		return nil, fmt.Errorf("bad config, cannot create L1 source: %w", err)
+	}
+	client = LimitRPC(client, config.MaxConcurrentRequests)
+	inner := &EthClient{
+		client:                  client,
+		maxBatchSize:            config.MaxRequestsPerBatch,
+		trustRPC:                config.TrustRPC,
+		mustBePostMerge:         config.MustBePostMerge,
+		provKind:                config.RPCProviderKind,
+		log:                     log,
+		receiptsCache:           caching.NewLRUCache[common.Hash, *receiptsFetchingJob](metrics, "receipts", config.ReceiptsCacheSize),
+		transactionsCache:       caching.NewLRUCache[common.Hash, types.Transactions](metrics, "txs", config.TransactionsCacheSize),
+		headersCache:            caching.NewLRUCache[common.Hash, eth.BlockInfo](metrics, "headers", config.HeadersCacheSize),
+		payloadsCache:           caching.NewLRUCache[common.Hash, *eth.ExecutionPayload](metrics, "payloads", config.PayloadsCacheSize),
+		availableReceiptMethods: AvailableReceiptsFetchingMethods(config.RPCProviderKind),
+		lastMethodsReset:        time.Now(),
+		methodResetDuration:     config.MethodResetDuration,
+		rethDbPath:              config.RethDBPath,
+	}
+	return &PrefetchingEthClient{
+		inner: inner,
+	}, nil
+}
+
+func (p *PrefetchingEthClient) SubscribeNewHead(ctx context.Context, ch chan<- *types.Header) (ethereum.Subscription, error) {
+	return p.inner.SubscribeNewHead(ctx, ch)
+}
+
+func (p *PrefetchingEthClient) ChainID(ctx context.Context) (*big.Int, error) {
+	return p.inner.ChainID(ctx)
+}
+
+func (p *PrefetchingEthClient) InfoByHash(ctx context.Context, hash common.Hash) (eth.BlockInfo, error) {
+	return p.inner.InfoByHash(ctx, hash)
+}
+
+func (p *PrefetchingEthClient) InfoByNumber(ctx context.Context, number uint64) (eth.BlockInfo, error) {
+	return p.inner.InfoByNumber(ctx, number)
+}
```

Comment:
```
p.inner.InfoByNumber(ctx, number)
for i in range(window):
    go p.inner.InfoByNumber(ctx,number + i)
```
The tests would then need to change because this hits the RPC a bunch of extra times